core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1525,14 +1516,102 @@ private Scope createScopeForView(
                             false))
                     .collect(toImmutableList());
 
-            analysis.addRelationCoercion(table, outputFields.stream().map(Field::getType).toArray(Type[]::new));
-
-            analyzeFiltersAndMasks(table, name, Optional.empty(), outputFields, session.getIdentity().getUser());
+            if (storageTable.isPresent()) {
+                // use storage table output fields
+                outputFields = analyzeStorageTable(table, outputFields, storageTable.get());
```

Comment:
The `outputFields` should be the ones derived from the materialized view definition. If there's a storage table, we should ensure they are compatible and add a coercion if necessary.

So, this should look closer to:

```java
            analyzeFiltersAndMasks(table, name, storageTable, outputFields, session.getIdentity().getUser());
            analysis.addRelationCoercion(table, outputFields.stream().map(Field::getType).toArray(Type[]::new));
            outputFields.forEach(field -> analysis.addSourceColumns(field, ImmutableSet.of(new SourceColumn(name, field.getName().orElseThrow()))));

            if (storageTable.isPresent()) {
                analyzeStorageTable(table, outputFields, storageTable.get());
            }
            else {
                analysis.registerNamedQuery(table, query);
            }

            return createAndAssignScope(table, scope, outputFields);
```

Reply:
> The outputFields should be the ones derived from the materialized view definition

They can't be. `analyzeStorageTable` calls `analyzeTableOutputFields` which registers fields (`analyzer.Analysis#setColumn`) with proper `ColumnHandles`. Since `Analysis#columns` keys are compared by `Field` identity, we have to register `storageTable` fields.

Then we validate `storageTable` compatibility.

I will improve the comment.

> we should ensure they are compatible and add a coercion if necessary.

Coercion check is already performed in `checkViewStaleness` above. We don't have to register it since we are using storage table (if it's present) that has matching column types.

Since storage table is strictly related to MV defined column types, we don't need to support coercion from storage table to MV column types.

Reply:
> Storage table is strictly related to MV column types (when it was created), therefore we don't need to support coercion from storage table columns to MV column types.

Technically, it could be an existing table. We haven't implemented that yet, but the idea was that someone might create a MV specify an existing storage table in the CREATE MV statement.

Reply:
> Technically, it could be an existing table. We haven't implemented that yet, but the idea was that someone might create a MV specify an existing storage table in the CREATE MV statement.

I propose we do this as follow-up. This PR is mostly about analyzing fresh MVs and enforcing security checks.